nncf/quantization/algorithms/accuracy_control/ranker.py

```diff
@@ -0,0 +1,301 @@
+        :param data_items: Data items.
+        :return: A list that contains logits for each item from the dataset.
+        """
+        engine = EngineFactory.create(model)
+        outputs = [
+            engine.infer(data_item) for data_item in data_items
+        ]
+        return outputs
+
+    def _calculate_ranking_score(self,
+                                 modified_model,
+                                 ranking_data_items: Iterable[Any],
+                                 ranking_subset_indices: List[int]) -> float:
+        approx_values_subset = self._collect_values_for_each_item(modified_model, ranking_data_items)
+        ref_values_subset = (self._ref_values[i] for i in ranking_subset_indices)
+        errors = [
+            self._ranking_fn(a, b) for a, b in zip(ref_values_subset, approx_values_subset)
+        ]
+        ranking_score = sum(errors) / len(errors)
+        return ranking_score
+
+
+class MetricBasedRanker(Ranker):
+    """
+    Encapsulates logic to rank groups of quantizers based on differences in metric.
+    """
+
+    def __init__(self,
+                 ranking_subset_size: int,
+                 ranking_fn: Callable[[Any, Any], float],
+                 dataset: Dataset,
+                 algo_backend: AccuracyControlAlgoBackend,
+                 validation_fn: Callable[[Any, Iterable[Any]], float]):
+        """
+        :param ranking_subset_size: The number of data items that will be selected from
+            the dataset to rank groups of quantizers. The `len(dataset)` data items will
+            be selected if `ranking_subset_size` parameter is greater than the number of
+            elements in the dataset.
+        :param ranking_fn:
+        :param dataset: Dataset for the ranking process.
+        :param algo_backend: The `AccuracyControlAlgoBackend` algo backend.
+        :param validation_fn: A validation function to validate the model.
+            It should take two argumets:
+                - `model`: model to be validate.
+                - `validation_dataset`: dataset that provides data items to
+                validate the provided model.
+            The function should return the value of the metric with the following
+            meaning: A higher value corresponds to better performance of the model.
+        """
+        super().__init__(ranking_subset_size, ranking_fn, dataset, algo_backend)
+        self._validation_fn = validation_fn
+
+    def _get_data_items(self, indices: Optional[List[int]] = None) -> Iterable[Any]:
+        return self._dataset.get_data(indices)
+
+    def _collect_values_for_each_item(self, model, data_items: Iterable[Any]) -> List[Any]:
+        """
+        Calls `validation_fn` for each item from the `dataset` and returns collected metrics.
+
+        :param model: The model to be inferred.
+        :param data_items: Data items.
+        :return: A list that contains a metric for each item from the dataset.
+        """
+        metrics = [
```

Comment:
What should we do for the dataset-based metrics that can't be collected for the specific indices? For example `coco_precision` for the MSCOCO. Or do we hope that the user will only have such a metric that can be calculated per-sample?

Reply:
As far as I know, we should use the NMSE metric in this case.

Reply:
Was the behavior with such metric tested? Is it working fine?

Reply:
A test was added.